index.js

```diff
@@ -16,28 +16,84 @@ const collectionName = 'dogs'
 app.get('/dogs', (req, res) => {
     mongoClient.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true }, (err, client) => {
         if (err) {
-            return res.status(500).json({
-                success: false,
-                message: 'Error connecting to database.',
-                data: []
-            })
+            client.close()
+            return composeJSON(res, 500, false, 'Error connecting to database.', [])
         }
         let db = client.db(dbName)
         let collection = db.collection(collectionName)
         collection.find({}).toArray((err, docs) => {
             if (err) {
-                return res.status(500).json({
-                    success: false,
-                    message: 'Error getting data from database.',
-                    data: []
-                })
+                client.close()
+                return composeJSON(res, 500, false, 'Error getting data from database.', [])
+            } else {
+                client.close()
+                return composeJSON(res, 200, true, 'Data retrieved successfully', docs)
             }
-            res.status(200).json({
-                success: true,
-                message: 'Data has been successfully retrieved and sent.',
-                data: docs
+        })
+    })
+})
+
+app.post('/dogs/:id/wins', (req, res) => {
+    const id = req.params.id
+    const regex = RegExp('[0-9a-f]{24}')
+    if (regex.exec(id) === null) {
+        return composeJSON(res, 400, false, 'Invalid winner ID.', [])
+    }
+    mongoClient.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true }, (err, client) => {
+        if (err) {
+            client.close()
+            return composeJSON(res, 500, false, 'Error connecting to database.', [])
+        }
+        const db = client.db(dbName)
+        const collection = db.collection(collectionName)
+        try {
+            declareChampion(collection, id, function (result) {
+                if (result.success === true) {
+                    client.close()
+                    return composeJSON(res, 200, true, 'Dog victory recorded', [result.result.result])
+                } else {
+                    client.close()
+                    return composeJSON(res, 400, false, 'Not able to update records.', [])
+                }
             })
+        }
+        catch (err) {
             client.close()
-        })
+            return composeJSON(res, 500, false, 'Failure attempting to update records.', [])
+        }
     })
 })
+
+const declareChampion = function (collection, id, callback) {
+    const winnerID = ObjectId(id)
+    try {
+        collection.updateOne(
+            {"_id": winnerID},
+            {$inc: {"winCount": 1}},
+            function (err, result) {
+                if (result.modifiedCount === 0){
```

Comment:
I think we should flip this, and be strict checking on success rather than failure - there may be many other ways that it could not be exactly `0`. Also, you could set the `success` and `result` values to variables inside the if's, and then pass them into callback after.

Reply:
done